src/engine/IndexScan.cpp

```diff
@@ -595,75 +614,112 @@ struct IndexScan::SharedGeneratorState {
-    prefetchedValues.clear();
-    for (auto& value : copy) {
-      co_yield value;
-    }
-  }
+  using LoopControl = ad_utility::LoopControl<Result::IdTableVocabPair>;
+
+  auto range = ad_utility::InputRangeFromLoopControlGet{
+      [state = std::move(innerState)]() mutable {
+        // Handle UNDEF case: pass through remaining input
+        if (state->hasUndef()) {
+          if (!state->iterator_.has_value()) {
+            state->iterator_ = state->generator_.begin();
+          }
+          auto& it = state->iterator_.value();
+          if (it == state->generator_.end()) {
+            return LoopControl::makeBreak();
+          }
+          auto result = std::move(*it);
+          ++it;
+          return LoopControl::yieldValue(std::move(result));
+        }
+
+        auto& prefetched = state->prefetchedValues_;
+        while (true) {
+          if (prefetched.empty()) {
+            if (state->doneFetching_) {
+              return LoopControl::makeBreak();
+            }
+            state->fetch();
+            AD_CORRECTNESS_CHECK(!prefetched.empty() || state->doneFetching_);
+            if (prefetched.empty()) {
+              return LoopControl::makeBreak();
+            }
+          }
+
+          // Make a defensive copy of the values to avoid modification during
+          // iteration when yielding.
+          auto copy = std::move(prefetched);
+          prefetched.clear();
+
+          // Yield all values from the copy, then break to fetch more
+          return LoopControl::yieldAll(std::move(copy));
+        }
+      }};
+  return Result::LazyResult{std::move(range)};
 }
 
 // _____________________________________________________________________________
-Result::Generator IndexScan::createPrefilteredIndexScanSide(
+Result::LazyResult IndexScan::createPrefilteredIndexScanSide(
     std::shared_ptr<SharedGeneratorState> innerState) {
+  using LoopControl = ad_utility::LoopControl<Result::IdTableVocabPair>;
+
+  // Handle UNDEF case using LoopControl pattern
   if (innerState->hasUndef()) {
-    for (auto& pair : chunkedIndexScan()) {
-      co_yield pair;
-    }
-    co_return;
-  }
-  LazyScanMetadata metadata;
-  auto& pendingBlocks = innerState->pendingBlocks_;
-  while (true) {
-    if (pendingBlocks.empty()) {
-      if (innerState->doneFetching_) {
-        metadata.numBlocksAll_ = innerState->metaBlocks_.sizeBlockMetadata_;
-        updateRuntimeInfoForLazyScan(metadata);
-        co_return;
-      }
-      innerState->fetch();
-    }
-    auto scan = getLazyScan(std::move(pendingBlocks));
-    AD_CORRECTNESS_CHECK(pendingBlocks.empty());
-    for (IdTable& idTable : scan) {
-      co_yield {std::move(idTable), LocalVocab{}};
-    }
-    metadata.aggregate(scan.details());
-  }
+    auto range = ad_utility::InputRangeFromLoopControlGet{[this]() {
+      return LoopControl::breakWithYieldAll(chunkedIndexScan());
+    }};
+    return Result::LazyResult{std::move(range)};
```

Comment:
To have the same lazy behavior as bevore, this should be inside the `range` below, s.t. it isn't called eagerly. Then you also don't need all the infrastructure, but just a plain

```
if (state->hasUndef()) {
 return breakWithYiledAll(chunkedIndexScan())`
};
```

And you are done.